### PR TITLE
Remove NodeID field from ATX

### LINKED ISSUE #3216: atx: remove node id field from atx
Sorry for only getting around to looking at this after it was merged, but I think we should remove the `NodeID` from the ATX. It was only there when we needed to maintain an identity database - now that we don't, we should derive the public key from the signature and do all validations on that (e.g. that the PoST proof validates with this key and that the key we derived from the previous ATX matches). It can (probably should) be cached in the database along with the ATX, but not included in the gossip message. This will be consistent with all our other messages.

_Originally posted by @noamnelke in https://github.com/spacemeshos/go-spacemesh/issues/3198#issuecomment-1128749597_

### PR DESCRIPTION
**Pull request by fasmat**
## Motivation
Closes #3216

## Changes
- Removed NodeId from ATX
- Derive NodeId from signature where needed
- Fix compile problems and tests

## Test Plan
- Unit tests via CI

## TODO
- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

**Comment by dshulyak** [comment]
there is a lot of unrelated changes in this pr

**Comment by fasmat** [comment]
> there is a lot of unrelated changes in this pr

There are a few changes I had to do because of updated dependencies. Do you want me to create a new PR with only those changes and rebase this PR after it is merged?

File: activation/handler.go [hunk 1]
```diff
@@ -153,7 +153,7 @@ func (h *Handler) SyntacticallyValidateAtx(ctx context.Context, atx *types.Activ
 	if err != nil {
 		return fmt.Errorf("cannot validate atx sig atx id %v err %v", atx.ShortString(), err)
 	}
-	if bytes.Compare(atx.NodeID[:], pub.Bytes()) != 0 {
+	if !bytes.Equal(atx.NodeID().ToBytes(), pub.Bytes()) {
```
Comment (dshulyak) [inline activation/handler.go]: i think we can drop this validation

**Comment by fasmat** [comment]
I removed all unrelated changes and will open a new PR fixing all complaints by `go-staticcheck`

EDIT: new PR for staticcheck fixes can be found here: https://github.com/spacemeshos/go-spacemesh/pull/3471

**Review by dshulyak — APPROVED** [review]

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors try

File: sql/atxs/atxs.go [hunk 2]
```diff
@@ -24,11 +24,15 @@ func Get(db sql.Executor, id types.ATXID) (atx *types.ActivationTx, err error) {
 		tickCount := uint64(stmt.ColumnInt64(2))
 		v.SetID(&id)
 		v.Verify(baseTickHeight, tickCount)
+		buf := make([]byte, len(types.NodeID{}))
```
Comment (countvonzero) [inline sql/atxs/atxs.go]: i think we should be consistent with other places by using `stmt.ColumnLen(3)` here

Comment (dshulyak) [reply sql/atxs/atxs.go]: i haven't noticed but you don't need that at all. 

```go
id := NodeID{}
stmt.ColumnBytes(3, id[:])
```

**Comment by dshulyak** [comment]
the reason why ci fails is because of

```
{"L":"WARN","T":"2022-08-22T09:30:09.367-0400","N":"5b27e.layerFetcher ","M":"failed fetching new ballots","node_id":"5b27e2d0f57c6a81ac16f1d205853e7f742851cab7884bedf81134f7893a0f03","module":"layerFetcher","sessionId":"618772c9-f5db-4404-b9c4-c90e943e8e6f","layer_id":15,"peer":"12D3KooWHgMeRUW2z2Q4x8re92yjef18D539f4JUVsF7AokSvdvG","num_ballots":1,"num_blocks":0,"ballot_ids":["dca568d894"],"errmsg":"syntactic-check ballot: fetch ballots: syntactic-check ballot: fetch referenced ATXs: proposal get ATXs: received syntactically invalid atx 2bd79145d4: previous atx belongs to different miner. atx.ID: 2bd79145d4, atx.NodeID: 214b9a69525c21730a6ac6d9eb1e2a8fee54eac2c6f14f6aacf1c428f042fba0, prevAtx.NodeID: 6a70574f406fb558c013f9b08a92c54751dafc46cc3d852ce1ee15c45192a3b2","name":"layerFetcher"}
```

previous atx should be from the same node. i wasn't able to understand why it started to fail with this change

**Comment by fasmat** [comment]
@countvonzero I cannot reproduce the failing test `TestShortTimeskew` locally. I'll look into what @dshulyak found.

**Comment by countvonzero** [comment]
> the reason why ci fails is because of
> 
> ```
> {"L":"WARN","T":"2022-08-22T09:30:09.367-0400","N":"5b27e.layerFetcher ","M":"failed fetching new ballots","node_id":"5b27e2d0f57c6a81ac16f1d205853e7f742851cab7884bedf81134f7893a0f03","module":"layerFetcher","sessionId":"618772c9-f5db-4404-b9c4-c90e943e8e6f","layer_id":15,"peer":"12D3KooWHgMeRUW2z2Q4x8re92yjef18D539f4JUVsF7AokSvdvG","num_ballots":1,"num_blocks":0,"ballot_ids":["dca568d894"],"errmsg":"syntactic-check ballot: fetch ballots: syntactic-check ballot: fetch referenced ATXs: proposal get ATXs: received syntactically invalid atx 2bd79145d4: previous atx belongs to different miner. atx.ID: 2bd79145d4, atx.NodeID: 214b9a69525c21730a6ac6d9eb1e2a8fee54eac2c6f14f6aacf1c428f042fba0, prevAtx.NodeID: 6a70574f406fb558c013f9b08a92c54751dafc46cc3d852ce1ee15c45192a3b2","name":"layerFetcher"}
> ```
> 
> previous atx should be from the same node. i wasn't able to understand why it started to fail with this change

so i tracked down this particular atx id `2bd79145d4`

```
{"L":"INFO","T":"2022-08-22T09:27:55.157-0400","N":"214b9.atxBuilder   ","M":"atx published","node_id":"214b9a69525c21730a6ac6d9eb1e2a8fee54eac2c6f14f6aacf1c428f042fba0","module":"atxBuilder","event":true,"challenge":"0xf1a2832ae4552766d983c742609263933026264572537de08ac5bf2984904133","id":"2bd79145d4","sender_id":"214b9a69525c21730a6ac6d9eb1e2a8fee54eac2c6f14f6aacf1c428f042fba0","prev_atx_id":"6967a7ce80","pos_atx_id":"ba002e0322","coinbase":"stest1qqqqqqp3qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqql50dsa","pub_layer_id":8,"epoch":2,"num_units":2,"sequence_number":1,"size":239,"name":"atxBuilder"}
```
the previous atx id is `6967a7ce80` and found taht two separate node id published the same atx 

```
{"L":"INFO","T":"2022-08-22T09:26:55.324-0400","N":"214b9.atxBuilder   ","M":"atx published","node_id":"214b9a69525c21730a6ac6d9eb1e2a8fee54eac2c6f14f6aacf1c428f042fba0","module":"atxBuilder","event":true,"nipost":"nonce: 6, indices: \"6008…\"","challenge":"0xc518895430d1ae9dc9aedf4b308ed55b94ab1b948d112a440db84a8992060270","id":"6967a7ce80","sender_id":"214b9a69525c21730a6ac6d9eb1e2a8fee54eac2c6f14f6aacf1c428f042fba0","prev_atx_id":"0000000000","pos_atx_id":"0000000000","coinbase":"stest1qqqqqqp3qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqql50dsa","pub_layer_id":4,"epoch":1,"num_units":2,"sequence_number":0,"size":249,"name":"atxBuilder"}
{"L":"INFO","T":"2022-08-22T09:26:55.075-0400","N":"6a705.atxBuilder   ","M":"atx published","node_id":"6a70574f406fb558c013f9b08a92c54751dafc46cc3d852ce1ee15c45192a3b2","module":"atxBuilder","event":true,"nipost":"nonce: 4, indices: \"6008…\"","challenge":"0xc518895430d1ae9dc9aedf4b308ed55b94ab1b948d112a440db84a8992060270","id":"6967a7ce80","sender_id":"6a70574f406fb558c013f9b08a92c54751dafc46cc3d852ce1ee15c45192a3b2","prev_atx_id":"0000000000","pos_atx_id":"0000000000","coinbase":"stest1qqqqqqp3qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqql50dsa","pub_layer_id":4,"epoch":1,"num_units":2,"sequence_number":0,"size":249,"name":"atxBuilder"}
```

@dshulyak 
@fasmat

btw, the namespace on CI for these logs is `test-zwar`
the queries i used are

{namespace="test-zwar"} | json | module="atxBuilder" | M="atx published" | id="2bd79145d4"
{namespace="test-zwar"} | json | module="atxBuilder" | M="atx published" | id="6967a7ce80"

**Comment by fasmat** [comment]
> the previous atx id is `6967a7ce80` and found taht two separate node id published the same atx

this gives me an idea! Maybe the reason is a similar one as for the other failing test we fixed already? ATX ID calculation used to include the Node ID, but now doesn't any more:

If both ATXs are identical (= reference the same previous ATX and have the same metadata) the ATX IDs are identical even if produced by two different nodes and signed by two different keys.

**Comment by fasmat** [comment]
Specifically I mean here: https://github.com/spacemeshos/go-spacemesh/blob/fix-%233216/common/types/activation.go#L288

and here https://github.com/spacemeshos/go-spacemesh/blob/fix-%233216/common/types/hashes.go#L177

Essentially only these 3 fields are used to derive the ATX ID:

https://github.com/spacemeshos/go-spacemesh/blob/fix-%233216/common/types/activation.go#L104-L106

I'm just not sure if it's just the test that causes these to be the same or if this is actually an oversight that needs fixing.

**Comment by countvonzero** [comment]
> this gives me an idea! Maybe the reason is a similar one as for the other failing test we fixed already? ATX ID calculation used to include the Node ID, but now doesn't any more:

what's wrong to me is that the two nodes should have different challenge. but in the ATX they each build, they had the same challenge `0xc518895430d1ae9dc9aedf4b308ed55b94ab1b948d112a440db84a8992060270`

**Comment by countvonzero** [comment]
> > this gives me an idea! Maybe the reason is a similar one as for the other failing test we fixed already? ATX ID calculation used to include the Node ID, but now doesn't any more:
> 
> what's wrong to me is that the two nodes should have different challenge. but in the ATX they each build, they had the same challenge `0xc518895430d1ae9dc9aedf4b308ed55b94ab1b948d112a440db84a8992060270`

but i am not completely sure. @dshulyak do you know?

**Comment by countvonzero** [comment]
> > > this gives me an idea! Maybe the reason is a similar one as for the other failing test we fixed already? ATX ID calculation used to include the Node ID, but now doesn't any more:
> > 
> > 
> > what's wrong to me is that the two nodes should have different challenge. but in the ATX they each build, they had the same challenge `0xc518895430d1ae9dc9aedf4b308ed55b94ab1b948d112a440db84a8992060270`
> 
> but i am not completely sure. @dshulyak do you know?

regardless, the two axt have different nonce. why do they still hash to the same thing?

**Comment by fasmat** [comment]
> regardless, the two axt have different nonce. why do they still hash to the same thing?

I don't think the nonce is used for ID calculation. Which field encodes the nonce in an ATX?

If it's `ActivationTx.InnerActivationTx.InitialPost.Nonce` or `ActivationTx.InnerActivationTx.NIPost.Post.Nonce` then it is not part of `CalcAndSetID` (only `ActivationTx.InnerActivationTx.ActivationTxHeader` is).

**Comment by dshulyak** [comment]
I found it very surprising that we computed atx id from atx header and not the whole structure (including signature). I guess it worked before (since node id was in the header), but this is really weird optimization. So i think we should change it to compute id using all data.

Regarding post (not post data) - i think it is not bound to the specific id, so thats why challenge might be the same.

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors try

**Review by dshulyak — CHANGES_REQUESTED** [review]
waiting for @noamnelke to clarify if we should rely on InitialPostIndices to avoid collision, or should we hash signature into atx id

**Comment by noamnelke** [comment]
@fasmat can you please help me validate my suspicion about the cause of the collisions?

I suspect that we have an issue with the difficulty calculation, causing a bias towards small index values. Can you please add a log for the actual indexes and try to find a collision? If you send me the actual colliding indexes I can test my hypothesis.

I just looked at the default parameters and I was a little surprised that `k1` is set to 2000... This means that with a default space unit of 2048, a single space unit smesher will basically only determine which 48 of their 2048 possible indexes **not** to include. The number of possible combinations is still huge, so we still need a bug or other issue to explain this.

I think that we can safely set this value to something MUCH smaller. Even if we set `k1` to 10, with a 256GiB space unit, the space of possibilities is immense (and the size of the proof would be tiny). Anyway, this is a separate discussion, not for right now.

**Pull request closed without merging**